csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,30 +15,66 @@ namespace nvfuser::preseg_passes {
 
 namespace {
 
+// returns allocation order to propagate adjusted for `tv`
+// Right now it merely removes all the entry corresponding to reduction iter
+// domains in tv,
+AllocationOrder adjustAllocationOrder(
+    const TensorView* tv,
+    const AllocationOrder& alloc_order) {
+  AllocationOrder ret = alloc_order;
+
+  int64_t tv_rank = static_cast<int64_t>(tv->nDims());
+  auto& rf_dom = tv->getMaybeRFactorDomain();
+  std::stack<int64_t> idx_stack;
+  // we scan all rfactor domain of tv and remove all reduction iter domain entry
+  // in `ret`.
+  for (auto idx : c10::irange(tv_rank)) {
+    if (rf_dom[idx]->isReduction()) {
+      // we push indices of each reduction iter domain into `idx_stack` in
+      // ascending order
+      idx_stack.push(idx);
+      for (auto i = ret.begin(); i != ret.end(); i++) {
+        if (*i == idx) {
+          ret.erase(i);
+          break;
+        }
+      }
+    }
+  }
+  // for any removed entry, we need to compensate the returned allocation order
+  // by decrement indices higher than the removed entry.
+  while (!idx_stack.empty()) {
+    auto idx = idx_stack.top();
+    idx_stack.pop();
+    for (auto i = ret.begin(); i != ret.end(); i++) {
+      if (*i > idx) {
+        --(*i);
+      }
+    }
+  }
```

Comment:
It will take me or any reader more than 10 minutes to review this :) Wdyt about the following alternative: 

1. construct the allocation domain from allocation order, e.g. `[r1,i2,i3,i0]` for the example in the PR description. 
2. map that `vector<IterDomain*>` to the consumer via PairwiseRootMap. I believe `r1` automatically disappear, so you have `[i2,i3,i0]`. 
3. compute the allocation order as `computePermutation(consumer's rfactor, [i2,i3,i0])`. 

Although this converts between allocation order and allocation domain, it makes the intent of filtering out reduction dimensions clearer, and avoids the index/offset calculation.

Reply:
> map that vector<IterDomain*> to the consumer via PairwiseRootMap.

I'm slightly uncomfortable with this change here, because it bind the logic here as a `producer->consumer` mapping. Which makes it a bit cumbersome to wield.

But I see your point of making the logic easier to read.

I'll refactor it with PairwiseRootDomainMap and remove the util function. (since broadcast propagation also uses it but needs a slightly different behavior.)